File: javascript/CsiUnits.js
/* CsiUnits.js

   Copyright (C) 2018, 2019 Campbell Scientific, Inc.

   Written by: Jon Trauntvein 
   Date Begun: Tuesday 27 November 2018
   Last Change: Friday 17 May 2019
   Last Commit: $Date: 2019-05-17 12:18:37 -0600 (Fri, 17 May 2019) $
   Last Changed by: $Author: jon $

*/


/**
 * Defines an object that apecifies the names of source and destination units as well as a function
 * that will convert a number based on the source and dest units.
 *
 * @param {string} source_unit Specifies the name of the source unit.
 *
 * @param {string} dest_unit Specifis the name of the destination unit.
  */
function CsiUnitConverterBase(source_unit, dest_unit)
{
   if(source_unit !== undefined && dest_unit !== undefined)
   {
      this.source_unit = source_unit;
      this.dest_unit = dest_unit;
   }
}


/**
 * @return {number} Returns the value parameter with the unit conversion function applied.
 *
 * @param {number} Specifies the value to convert.
 */
CsiUnitConverterBase.prototype.convert = function(value)
{
   return value;
};



/**
 * Defines an object that performs a linear transformation (multiplier and offset) for
 * unit conversion.
 *
 * @param source_unit {string} Specifies the name of the source unit.
 *
 * @param dest_unit {string} Specifies the name of the dest unit.
 *
 * @param invert {boolean} Set to true if the conversion should be inverted.
 *
 * @param multiplier {number} Specifies the multiplier that should be applied.
 *
 * @param offset {number} Specifies the offset that should be applied.
 */
function CsiUnitConverterLinear(source_unit, dest_unit, invert, multiplier, offset)
{
   if(source_unit !== undefined)
   {
      CsiUnitConverterBase.call(this, source_unit, dest_unit);
      this.invert = invert;
      this.multiplier = multiplier;
      this.offset = offset;
   }
}
CsiUnitConverterLinear.prototype = new CsiUnitConverterBase();


CsiUnitConverterLinear.prototype.convert = function(value)
{
   var rtn;
   if(this.invert)
      rtn = (value - this.offset) / this.multiplier;
   else
      rtn = value * this.multiplier + this.offset;
   return rtn;
};


/**
 * Defines a JavaScript component that can be used to translate between similar units which are identified
 * by their case sensitive NIST abbreviations.
 */
function CsiUnits()
{
   var standard_conversions;
   var units = this;
   if(this instanceof CsiUnits)
   {
      this.converters = {};
      standard_conversions = [
         { source_unit: "m", dest_unit: "km", multiplier: 1E-3, offset: 0 },
         { source_unit: "m", dest_unit: "cm", multiplier: 1E2, offset: 0 },
         { source_unit: "m", dest_unit: "mm", multiplier: 1E3, offset: 0 },
         { source_unit: "m", dest_unit: "\u03bcm", multiplier: 1E6, offset: 0 },
         { source_unit: "m", dest_unit: "nm", multiplier: 1E9, offset: 0 },
         { source_unit: "m", dest_unit: "ft", multiplier: 3.28084, offset: 0 },
         { source_unit: "ft", dest_unit: "in", multiplier: 12.0, offset: 0 },
         { source_unit: "ft", dest_unit: "mi", multiplier: 0.000189394, offset: 0 },
         { source_unit: "ft\u00b2", dest_unit: "gal", multiplier: 0.133681, offset: 0 },
         { source_unit: "m\u00b2", dest_unit: "km\u00b2", multiplier: 1E-6, offset: 0 },
         { source_unit: "m\u00b2", dest_unit: "ha", multiplier: 1E-4, offset: 0 },
         { source_unit: "m\u00b2", dest_unit: "ft\u00b2", multiplier: 10.7639, offset: 0 },
         { source_unit: "m\u00b3", dest_unit: "ft\u00b3", multiplier: 35.3147, offset: 0 },
         { source_unit: "m\u00b3", dest_unit: "L", multiplier: 1E3, offset: 0 },
         { source_unit: "L", dest_unit: "mL", multiplier: 1E3, offset: 0 },
         { source_unit: "L", dest_unit: "\u2c98L", multiplier: 1E6, offset: 0 },
         { source_unit: "L", dest_unit: "gal", multiplier: 0.26, offset: 0 },
         { source_unit: "gal", dest_unit: "qt", multiplier: 4, offset: 0 },
         { source_unit: "gal", dest_unit: "pt", multiplier: 8, offset: 0 },
         { source_unit: "gal", dest_unit: "cup", multiplier: 15.7725, offset: 0 },
         { source_unit: "gal", dest_unit: "oz", multiplier: 128, offset: 0 },
         { source_unit: "g", dest_unit: "kg", multiplier: 1E-3, offset: 0 },
         { source_unit: "g", dest_unit: "mg", multiplier: 1E3, offset: 0 },
         { source_unit: "g", dest_unit: "\u03bcg", multiplier: 1E-6, offset: 0 },
         { source_unit: "m/s", dest_unit: "km/h", multiplier: 3.6, offset: 0 },
         { source_unit: "m/s", dest_unit: "mph", multiplier: 2.23694, offset: 0 },
         { source_unit: "mph", dest_unit: "MPH", multiplier: 1, offset: 0 },
         { source_unit: "mph", dest_unit: "mi/h", multiplier: 1, offset: 0 },
         { source_unit: "m/s", dest_unit: "c", multiplier: 299792458, offset: 0 },
         { source_unit: "mph", dest_unit: "ft/s", multiplier: 1.46667, offset: 0 },
         { source_unit: "degC", dest_unit: "\u00b0C", multiplier: 1, offset: 0 },
         { source_unit: "\u00b0C", dest_unit: "\u00b0F", multiplier: 1.8, offset: 32 },
         { source_unit: "degF", dest_unit: "\u00b0F", multiplier: 1, offset: 0 },
         { source_unit: "\u00b0C", dest_unit: "K", multiplier: 1.0, offset: 273.15 },
         { source_unit: "s", dest_unit: "\u03bcs", multiplier: 1E6, offset: 0 },
         { source_unit: "s", dest_unit: "ms", multiplier: 1E3, offset: 0 },
         { source_unit: "s", dest_unit: "ns", multiplier: 1E9, offset: 0 },
         { source_unit: "s", dest_unit: "min", multiplier: 1.0/60.0, offset: 0 },
         { source_unit: "s", dest_unit: "hr", multiplier: 1.0/3600.0, offset: 0 },
         { source_unit: "s", dest_unit: "day", multiplier: 1.0/86400.0, offset: 0 },
         { source_unit: "s", dest_unit: "week", multiplier: 1.0/604800.0, offset: 0 },
         { source_unit: "kPa", dest_unit: "Pa", multiplier: 1E3, offset: 0 },
         { source_unit: "kPa", dest_unit: "b", multiplier: 0.01, offset: 0 },
         { source_unit: "kPa", dest_unit: "mmHg", multiplier: 7.500616827, offset: 0 },
         { source_unit: "kPa", dest_unit: "psi", multiplier: 0.145038, offset: 0 },
         { source_unit: "kPa", dest_unit: "atm", multiplier: 0.00986923, offset: 0 },
         { source_unit: "kPa", dest_unit: "inHg", multiplier: 0.2953, offset: 0 },
         { source_unit: "KPa", dest_unit: "kPa", multiplier: 1, offset: 0 },
         { source_unit: "W/m^2", dest_unit: "W/m\u00b2", multiplier: 1, offset: 0 },
         { source_unit: "C", dest_unit: "\u00b0C", multiplier: 1, offset: 0 }
      ]
      standard_conversions.forEach(function(conversion) {
         units.add_converter(new CsiUnitConverterLinear(conversion.source_unit, conversion.dest_unit, false, conversion.multiplier, conversion.offset));
         units.add_converter(new CsiUnitConverterLinear(conversion.dest_unit, conversion.source_unit, true, conversion.multiplier, conversion.offset));
      });
   }
}


/**
 * Adds the specified converter object to the set of supported conversions.
 *
 ' @param {CsiConverterBase} converter Specifies the object that will do the conversion.
 */
CsiUnits.prototype.add_converter = function(converter)
{
   var unit_converter;
   if(this.converters.hasOwnProperty(converter.source_unit))
      unit_converter = this.converters[converter.source_unit];
   else
   {
      unit_converter = { };
      this.converters[converter.source_unit] = unit_converter;
   }
   unit_converter[converter.dest_unit] = converter;
};


/**
 * @return {number} Returns the value after the required set of units conversions have been applied.
 *
 * @param {number} val Specifies the value to convert.
 *
 * @param {string} source_units Specifies the source units for the value.
 *
 * @param {string} dest_units Specifies the destination units for the value.
 */
CsiUnits.prototype.convert = function(val, source_units, dest_units)
{
   var rtn = val;
   var selected_chain = null;
   var stack = [];
   var visited = new Set();
   var stack_elem;
   var current_source;
   var units = this;
   
   if(source_units != dest_units)
   {
      // there may not be a direct conversion from the source unit to the dest unit so we might have
      // to search recursively.  We will do this by using a stack and forming a chain of converter
      // objects that will be able to make the conversion.
      stack.push({unit: source_units, chain: [] });
      visited.add(source_units);
      while(stack.length > 0 && selected_chain === null)
      {
         stack_elem = stack.shift();
         if(this.converters.hasOwnProperty(stack_elem.unit))
         {
            current_source = this.converters[stack_elem.unit];
            if(current_source.hasOwnProperty(dest_units))
            {
               selected_chain = stack_elem.chain;
               selected_chain.push(current_source[dest_units]);
               break;
            }
            else
            {
               $.each(current_source, function(key, converter) {
                  var next_elem;
                  if(!visited.has(key))
                  {
                     visited.add(key);
                     next_elem = { unit: key, chain: stack_elem.chain.slice() };
                     next_elem.chain.push(converter);
                     stack.push(next_elem);
                  }
               });
            }
         }
      }
   }

   // we can now apply the chain of converters.
   if(selected_chain)
   {
      selected_chain.forEach(function(converter) {
         rtn = converter.convert(rtn);
      });
   }
   return rtn;
};

File: coratools/Csi.Units.cpp
/* Csi.Units.cpp

   Copyright (C) 2018, 2019 Campbell Scientific, Inc.

   Written by: Jon Trauntvein 
   Date Begun: Friday 16 November 2018
   Last Change: Wednesday 13 March 2019
   Last Commit: $Date: 2019-03-13 13:25:17 -0600 (Wed, 13 Mar 2019) $
   Last Changed by: $Author: jon $

*/

#pragma hdrstop               // stop creation of precompiled header
#include "Csi.Units.h"
#include <deque>
#include <set>
#include <limits>


namespace Csi
{
   namespace Units
   {
      namespace
      {
         struct LinearConverter: public ConverterBase
         {
            double const multiplier;
            double const offset;
            bool const inverse;
            
            LinearConverter(
               StrUni const &source,
               StrUni const &dest,
               bool const inverse_,
               double multiplier_,
               double offset_):
               ConverterBase(source, dest),
               multiplier(multiplier_),
               offset(offset_),
               inverse(inverse_)
            { }

            virtual double convert(double val) const
            {
               double rtn = std::numeric_limits<double>::quiet_NaN();
               if(!inverse)
                  rtn = val * multiplier + offset;
               else
                  rtn = (val - offset) / multiplier;
               return rtn;
            }
         };


         /**
          * Defines the set of linear conversions that we can support.  This list will be terminated
          * when both the source and dest strings are empty.
          */
         struct supported_linear_conversion_type
         {
            char const *source;
            char const *dest;
            double multiplier;
            double offset;
         } supported_linear_conversions[] =
         {
            { "m", "km", 1E-3, 0 },
            { "m", "cm", 1E2, 0 },
            { "m", "mm", 1E3, 0 },
            { "m", "\xce\xbcm", 1E6, 0 },
            { "m", "nm", 1E9, 0 },
            { "m", "ft", 3.28084, 0 },
            { "ft", "in", 12.0, 0 },
            { "ft", "yd", 3, 0 },
            { "ft", "mi", 0.000189394, 0 },
            { "ft\xc2\xb3", "gal", 0.133681, 0 },
            { "m\xc2\xb2", "km\xc2\xb2", 1E-6, 0 },
            { "m\xc2\xb2", "ha", 1E-4, 0 },
            { "m\xc2\xb2", "ft\xc2\xb2", 10.7639, 0 },
            { "m\xc2\xb3", "ft\xc2\xb3", 35.3147, 0 },
            { "m\xc2\xb3", "L", 1E3, 0 },
            { "m\xc2\xb3", "mL", 1E6, 0 },
            { "m\xc2\xb3", "\xe2\xbcL", 1E6, 0 },
            { "L", "gal", 0.26, 0 },
            { "gal", "qt", 4, 0 },
            { "gal", "pt", 8, 0 },
            { "gal", "cup", 15.7725, 0 },
            { "gal", "oz", 128, 0 },
            { "g", "kg", 1E-3, 0 },
            { "g", "mg", 1E3, 0 },
            { "g", "\xce\xbcg", 1E6, 0 },
            { "m/s", "km/h", 3.6, 0 },
            { "m/s", "MPH", 2.23694, 0 },
            { "MPH", "mph", 1, 0 },
            { "MPH", "mi/h", 1, 0 },
            { "m/s", "c", 299792458, 0 },
            { "MPH", "ft/s", 1.46667, 0 },
            { "degC", "\xc2\xb0\x43", 1, 0 },
            { "C", "\xc2\xb0\x43", 1, 0 },
            { "degF", "\xc2\xb0\x46", 1, 0 },
            { "\xc2\xb0\x43", "\xc2\xb0\x46", 1.8, 32 },
            { "\xc2\xb0\x43", "K", 1.0, 273.15 },
            { "s", "ms", 1E3, 0 },
            { "s", "\xce\xbcs", 1E6, 0 },
            { "s", "ns", 1E9, 0 },
            { "s", "min", 1.0/60.0, 0 },
            { "s", "hr", 1.0/3600.0, 0 },
            { "s", "day", 1.0/86400.0, 0 },
            { "s", "week", 1.0/604800.0, 0 },
            { "kPa", "Pa", 1E3, 0 },
            { "kPa", "b", 0.01, 0 },
            { "kPa", "mmHg", 7.500616827, 0 },
            { "kPa", "psi", 0.145038, 0 },
            { "kPa", "atm", 0.00986923, 0 },
            { "kPa", "inHg", 0.2953, 0 },
            { "W/m^2", "W/m\xc2\xb2", 1, 0 },
            { 0, 0, 0, 0 }
         };
      };
      


      Converter::Converter()
      {
         for(size_t i = 0; supported_linear_conversions[i].source != 0; ++i)
         {
            add_converter(
               new LinearConverter(
                  supported_linear_conversions[i].source,
                  supported_linear_conversions[i].dest,
                  false,
                  supported_linear_conversions[i].multiplier,
                  supported_linear_conversions[i].offset));
            add_converter(
               new LinearConverter(
                  supported_linear_conversions[i].dest,
                  supported_linear_conversions[i].source,
                  true,
                  supported_linear_conversions[i].multiplier,
                  supported_linear_conversions[i].offset));
         }
      } // constructor


      Converter::~Converter()
      { converters.clear(); }


      double Converter::convert(double val, StrUni const &source_units, StrUni const &dest_units) const
      {
         typedef std::deque<converter_handle> chain_type;
         typedef std::pair<StrUni, chain_type> stack_elem_type;
         typedef std::deque<stack_elem_type> stack_type;
         typedef std::set<StrUni, ConverterHelpers::compare_no_case> visited_units_type;
         double rtn(val);
         chain_type selected_chain;
         
         if(source_units.compare(dest_units, true) != 0)
         {
            // there may not be a direct conversion from the source to the destination units so we
            // might have to perform a recursive search.  We will do this by using a stack and building
            // a chain of converters that can be called to perform the conversion
            stack_type stack;
            visited_units_type visited_units;

            stack.push_back(stack_elem_type(source_units, chain_type()));
            visited_units.insert(source_units);
            while(!stack.empty() && selected_chain.empty())
            {
               stack_elem_type elem(stack.front());
               converters_type::const_iterator ci(converters.find(elem.first));
               stack.pop_front();
               if(ci != converters.end())
               {
                  // there is a chance we can find the direct conversion on this step using the
                  // find() method.  if that doesn't work, will will have to push all of the
                  // previously unvisited conversions at this level onto the stack.
                  unit_converters_type::const_iterator uci(ci->second.find(dest_units));
                  if(uci != ci->second.end())
                  {
                     selected_chain = elem.second;
                     selected_chain.push_back(uci->second);
                     break;
                  }
                  else
                  {
                     for(uci = ci->second.begin(); uci != ci->second.end(); ++uci)
                     {
                        if(visited_units.find(uci->first) == visited_units.end())
                        {
                           stack_elem_type next_elem(uci->first, elem.second);
                           next_elem.second.push_back(uci->second);
                           stack.push_back(next_elem);
                           visited_units.insert(uci->first);
                        }
                     }
                  }
               }
            }
         }

         // we will now run through the chain of conversions needed.
         for(chain_type::const_iterator ci = selected_chain.begin(); ci != selected_chain.end(); ++ci)
            rtn = (*ci)->convert(rtn);
         return rtn;
      } // convert
   };
};
